diff --git a/qreps_val/experiments/cartpole.py b/qreps_val/experiments/cartpole.py
index 6109976..89287dd 100644
--- a/qreps_val/experiments/cartpole.py
+++ b/qreps_val/experiments/cartpole.py
@@ -1,5 +1,11 @@
+import argparse
 import logging
+import random
 import sys, os
+import time
+import numpy as np
+
+import wandb
 
 sys.path.append("../")
 sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))
@@ -15,40 +21,83 @@ from qreps.algorithms.sampler import ExponentiatedGradientSampler
 from qreps.algorithms.sampler import BestResponseSampler
 from qreps.feature_functions import IdentityFeature
 from qreps.policies.qreps_policy import QREPSPolicy
+from qreps.policies.categorical_mlp import CategoricalMLP
 from qreps.utilities.trainer import Trainer
 from qreps.utilities.util import set_seed
-from qreps.valuefunctions import NNQFunction, DiscreteMLPCritic, ResNet
+from qreps.valuefunctions import NNQFunction, DiscreteMLPCritic
 
 for handler in logging.root.handlers[:]:
     logging.root.removeHandler(handler)
 
+
+def make_env(env_id, capture_video, run_name):
+    def thunk():
+        if capture_video:
+            env = gym.make(env_id, render_mode="rgb_array")
+            env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
+        else:
+            env = gym.make(env_id)
+        env = gym.wrappers.RecordEpisodeStatistics(env)
+        return env
+
+    return thunk()
+
 FORMAT = "[%(asctime)s]: %(message)s"
 logging.basicConfig(level=logging.INFO, format=FORMAT)
 
 SEED_OFFSET = 0
 
 qreps_config = {
-    "eta": 0.1,
-    "beta": 2e-2,
+    "env_id": "CartPole-v1",  
+    "track": True,
+    "eta": 0.01,
+    "beta": 8e-2,
     "saddle_point_steps": 300,
     "policy_opt_steps": 300,
     "discount": 0.99,
-    "device": "cuda" if torch.cuda.is_available() else "cpu",
+    "wandb_project_name": "qreps",
+    "wandb_entity": None,
+    "seed" : 14,
+    "exp_name": os.path.basename(__file__)[: -len(".py")],
+    "num_envs": 1,
+    "capture_video": True,
 }
 
 
+
 def train(config: dict):
+    config = argparse.Namespace(**config)
+    run_name = f"{config.env_id}__{config.exp_name}__{config.seed}__{int(time.time())}"
+    if config.track:
+        wandb.init(
+                project=config.wandb_project_name,
+                entity=config.wandb_entity,
+                sync_tensorboard=True,
+                config=vars(config),
+                name=run_name,
+                monitor_gym=True,
+                save_code=True,
+            )
+    writer = SummaryWriter(f"runs/{run_name}")
+
+    random.seed(config.seed)
+    np.random.seed(config.seed)
+    torch.manual_seed(config.seed)
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    env = make_env(config.env_id, config.capture_video, run_name)
+
+    # env = gym.make("CartPole-v1", render_mode="rgb_array")
+    # env = gym.wrappers.RecordVideo(env, f"videos/{'cartpole'}")
 
-    env = gym.make("CartPole-v1", render_mode="rgb_array")
-    env = gym.wrappers.RecordVideo(env, f"videos/{'cartpole'}")
     num_obs = env.observation_space.shape[0]
     num_act = env.action_space.n
-    device = config["device"]
 
     q_function = NNQFunction(
         obs_dim=num_obs, act_dim=num_act, feature_fn=IdentityFeature()
     )
-    policy = QREPSPolicy(q_function=q_function, temp=config["eta"])
+    # policy = CategoricalMLP(num_obs, num_act)
+    policy = QREPSPolicy(q_function, temp=config.eta)
     writer = SummaryWriter()
 
     agent = QREPS(
@@ -58,14 +107,16 @@ def train(config: dict):
         learner=torch.optim.Adam,
         sampler=BestResponseSampler,
         optimize_policy=False,
-        reward_transformer=lambda r: r / 25000 ,
-        **config,
-    )
+        policy_lr=1e-3,
+        reward_transformer=lambda r: r / 1000,
+        device = device,
+        **vars(config))
 
     trainer = Trainer()
     trainer.setup(agent, env)
-    trainer.train(num_iterations=30, max_steps=500, number_rollouts=50)
-
+    trainer.train(num_iterations=30, max_steps=200, number_rollouts=5)
+    env.close()
+    writer.close()
 
 train(qreps_config)
 
diff --git a/qreps_val/qreps/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/__pycache__/__init__.cpython-311.pyc
index 1fae256..c644957 100644
Binary files a/qreps_val/qreps/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/algorithms/__pycache__/__init__.cpython-311.pyc
index 7319b3d..ae3b122 100644
Binary files a/qreps_val/qreps/algorithms/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/algorithms/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/__pycache__/abstract_algorithm.cpython-311.pyc b/qreps_val/qreps/algorithms/__pycache__/abstract_algorithm.cpython-311.pyc
index 78646c4..87a324b 100644
Binary files a/qreps_val/qreps/algorithms/__pycache__/abstract_algorithm.cpython-311.pyc and b/qreps_val/qreps/algorithms/__pycache__/abstract_algorithm.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/__pycache__/qreps.cpython-311.pyc b/qreps_val/qreps/algorithms/__pycache__/qreps.cpython-311.pyc
index 780597c..acbe541 100644
Binary files a/qreps_val/qreps/algorithms/__pycache__/qreps.cpython-311.pyc and b/qreps_val/qreps/algorithms/__pycache__/qreps.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/__pycache__/saddle_qreps.cpython-311.pyc b/qreps_val/qreps/algorithms/__pycache__/saddle_qreps.cpython-311.pyc
index 12d203b..03610c8 100644
Binary files a/qreps_val/qreps/algorithms/__pycache__/saddle_qreps.cpython-311.pyc and b/qreps_val/qreps/algorithms/__pycache__/saddle_qreps.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/sampler/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/algorithms/sampler/__pycache__/__init__.cpython-311.pyc
index f0f1755..5f69128 100644
Binary files a/qreps_val/qreps/algorithms/sampler/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/algorithms/sampler/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/sampler/__pycache__/abstract_sampler.cpython-311.pyc b/qreps_val/qreps/algorithms/sampler/__pycache__/abstract_sampler.cpython-311.pyc
index ea788da..c34ca0d 100644
Binary files a/qreps_val/qreps/algorithms/sampler/__pycache__/abstract_sampler.cpython-311.pyc and b/qreps_val/qreps/algorithms/sampler/__pycache__/abstract_sampler.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/sampler/__pycache__/best_response.cpython-311.pyc b/qreps_val/qreps/algorithms/sampler/__pycache__/best_response.cpython-311.pyc
index d52cd04..890287e 100644
Binary files a/qreps_val/qreps/algorithms/sampler/__pycache__/best_response.cpython-311.pyc and b/qreps_val/qreps/algorithms/sampler/__pycache__/best_response.cpython-311.pyc differ
diff --git a/qreps_val/qreps/algorithms/sampler/__pycache__/exponentiated_gradient.cpython-311.pyc b/qreps_val/qreps/algorithms/sampler/__pycache__/exponentiated_gradient.cpython-311.pyc
index b49485d..a09976e 100644
Binary files a/qreps_val/qreps/algorithms/sampler/__pycache__/exponentiated_gradient.cpython-311.pyc and b/qreps_val/qreps/algorithms/sampler/__pycache__/exponentiated_gradient.cpython-311.pyc differ
diff --git a/qreps_val/qreps/feature_functions/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/feature_functions/__pycache__/__init__.cpython-311.pyc
index 171a0e3..66b00f3 100644
Binary files a/qreps_val/qreps/feature_functions/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/feature_functions/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/feature_functions/__pycache__/abstract_feature_function.cpython-311.pyc b/qreps_val/qreps/feature_functions/__pycache__/abstract_feature_function.cpython-311.pyc
index d8262b1..43b4892 100644
Binary files a/qreps_val/qreps/feature_functions/__pycache__/abstract_feature_function.cpython-311.pyc and b/qreps_val/qreps/feature_functions/__pycache__/abstract_feature_function.cpython-311.pyc differ
diff --git a/qreps_val/qreps/feature_functions/__pycache__/identity.cpython-311.pyc b/qreps_val/qreps/feature_functions/__pycache__/identity.cpython-311.pyc
index 5db7450..33c001f 100644
Binary files a/qreps_val/qreps/feature_functions/__pycache__/identity.cpython-311.pyc and b/qreps_val/qreps/feature_functions/__pycache__/identity.cpython-311.pyc differ
diff --git a/qreps_val/qreps/memory/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/memory/__pycache__/__init__.cpython-311.pyc
index c3423d0..24f574e 100644
Binary files a/qreps_val/qreps/memory/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/memory/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/memory/__pycache__/buffer.cpython-311.pyc b/qreps_val/qreps/memory/__pycache__/buffer.cpython-311.pyc
index f335b53..53f41b3 100644
Binary files a/qreps_val/qreps/memory/__pycache__/buffer.cpython-311.pyc and b/qreps_val/qreps/memory/__pycache__/buffer.cpython-311.pyc differ
diff --git a/qreps_val/qreps/memory/__pycache__/replay_buffer.cpython-311.pyc b/qreps_val/qreps/memory/__pycache__/replay_buffer.cpython-311.pyc
index 8f02400..8707cc5 100644
Binary files a/qreps_val/qreps/memory/__pycache__/replay_buffer.cpython-311.pyc and b/qreps_val/qreps/memory/__pycache__/replay_buffer.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/__init__.cpython-311.pyc
index 4c15c0f..e65f25a 100644
Binary files a/qreps_val/qreps/policies/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/abstract_q_policy.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/abstract_q_policy.cpython-311.pyc
index e13baaf..6181667 100644
Binary files a/qreps_val/qreps/policies/__pycache__/abstract_q_policy.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/abstract_q_policy.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/categorical_mlp.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/categorical_mlp.cpython-311.pyc
index 1789e98..be2a467 100644
Binary files a/qreps_val/qreps/policies/__pycache__/categorical_mlp.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/categorical_mlp.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/direct_set_policy.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/direct_set_policy.cpython-311.pyc
index d9c9fde..186ebed 100644
Binary files a/qreps_val/qreps/policies/__pycache__/direct_set_policy.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/direct_set_policy.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/gaussian_mlp.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/gaussian_mlp.cpython-311.pyc
index fbd9fb0..d219d2f 100644
Binary files a/qreps_val/qreps/policies/__pycache__/gaussian_mlp.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/gaussian_mlp.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/qreps_policy.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/qreps_policy.cpython-311.pyc
index 3e47485..86e14a3 100644
Binary files a/qreps_val/qreps/policies/__pycache__/qreps_policy.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/qreps_policy.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/stochastic_table.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/stochastic_table.cpython-311.pyc
index 829bf82..c008c73 100644
Binary files a/qreps_val/qreps/policies/__pycache__/stochastic_table.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/stochastic_table.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/stochasticpolicy.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/stochasticpolicy.cpython-311.pyc
index f70f352..15c2a60 100644
Binary files a/qreps_val/qreps/policies/__pycache__/stochasticpolicy.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/stochasticpolicy.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/__pycache__/value_function_policy.cpython-311.pyc b/qreps_val/qreps/policies/__pycache__/value_function_policy.cpython-311.pyc
index a22139a..2a09b03 100644
Binary files a/qreps_val/qreps/policies/__pycache__/value_function_policy.cpython-311.pyc and b/qreps_val/qreps/policies/__pycache__/value_function_policy.cpython-311.pyc differ
diff --git a/qreps_val/qreps/policies/categorical_mlp.py b/qreps_val/qreps/policies/categorical_mlp.py
index 78fd1a0..3bcdc90 100644
--- a/qreps_val/qreps/policies/categorical_mlp.py
+++ b/qreps_val/qreps/policies/categorical_mlp.py
@@ -1,20 +1,25 @@
+import numpy as np
 import torch
 import torch.nn as nn
 
 from .stochasticpolicy import StochasticPolicy
 
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
 
 class CategoricalMLP(StochasticPolicy, nn.Module):
     def __init__(self, obs_shape, act_shape, *args, **kwargs):
         super(CategoricalMLP, self).__init__(*args, **kwargs)
         self.model = nn.Sequential(
-            nn.Linear(obs_shape, 200),
+            (nn.Linear(obs_shape, 256)),
             nn.ReLU(),
-            nn.Linear(200, 200),
+            (nn.Linear(256, 256)),
             nn.ReLU(),
-            nn.Linear(200, 200),
+            (nn.Linear(256, 256)),
             nn.ReLU(),
-            nn.Linear(200, act_shape),
+            nn.Linear(256, act_shape),
         )
 
     def forward(self, x):
diff --git a/qreps_val/qreps/utilities/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/__init__.cpython-311.pyc
index 48ca167..5994583 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/utilities/__pycache__/elbe.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/elbe.cpython-311.pyc
index 5cb00bb..9ef5333 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/elbe.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/elbe.cpython-311.pyc differ
diff --git a/qreps_val/qreps/utilities/__pycache__/math.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/math.cpython-311.pyc
index 7e08416..e5f2e84 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/math.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/math.cpython-311.pyc differ
diff --git a/qreps_val/qreps/utilities/__pycache__/nn.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/nn.cpython-311.pyc
index 1313762..110e513 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/nn.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/nn.cpython-311.pyc differ
diff --git a/qreps_val/qreps/utilities/__pycache__/trainer.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/trainer.cpython-311.pyc
index 28c9a9f..9a411b1 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/trainer.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/trainer.cpython-311.pyc differ
diff --git a/qreps_val/qreps/utilities/__pycache__/util.cpython-311.pyc b/qreps_val/qreps/utilities/__pycache__/util.cpython-311.pyc
index da84d73..f2c15b9 100644
Binary files a/qreps_val/qreps/utilities/__pycache__/util.cpython-311.pyc and b/qreps_val/qreps/utilities/__pycache__/util.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/__pycache__/AccumulativeModule.cpython-311.pyc b/qreps_val/qreps/valuefunctions/__pycache__/AccumulativeModule.cpython-311.pyc
index dc52647..cd10828 100644
Binary files a/qreps_val/qreps/valuefunctions/__pycache__/AccumulativeModule.cpython-311.pyc and b/qreps_val/qreps/valuefunctions/__pycache__/AccumulativeModule.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/__pycache__/__init__.cpython-311.pyc b/qreps_val/qreps/valuefunctions/__pycache__/__init__.cpython-311.pyc
index 1a02abb..4d49db3 100644
Binary files a/qreps_val/qreps/valuefunctions/__pycache__/__init__.cpython-311.pyc and b/qreps_val/qreps/valuefunctions/__pycache__/__init__.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/__pycache__/integrated_q_function.cpython-311.pyc b/qreps_val/qreps/valuefunctions/__pycache__/integrated_q_function.cpython-311.pyc
index fd40e54..c6ab16c 100644
Binary files a/qreps_val/qreps/valuefunctions/__pycache__/integrated_q_function.cpython-311.pyc and b/qreps_val/qreps/valuefunctions/__pycache__/integrated_q_function.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/__pycache__/q_function.cpython-311.pyc b/qreps_val/qreps/valuefunctions/__pycache__/q_function.cpython-311.pyc
index a09abe4..aee8433 100644
Binary files a/qreps_val/qreps/valuefunctions/__pycache__/q_function.cpython-311.pyc and b/qreps_val/qreps/valuefunctions/__pycache__/q_function.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/__pycache__/value_functions.cpython-311.pyc b/qreps_val/qreps/valuefunctions/__pycache__/value_functions.cpython-311.pyc
index a8e2f68..1608c38 100644
Binary files a/qreps_val/qreps/valuefunctions/__pycache__/value_functions.cpython-311.pyc and b/qreps_val/qreps/valuefunctions/__pycache__/value_functions.cpython-311.pyc differ
diff --git a/qreps_val/qreps/valuefunctions/q_function.py b/qreps_val/qreps/valuefunctions/q_function.py
index cdad479..bc7d610 100644
--- a/qreps_val/qreps/valuefunctions/q_function.py
+++ b/qreps_val/qreps/valuefunctions/q_function.py
@@ -3,14 +3,13 @@ from functools import partial
 import torch.nn as nn
 import math, torch
 import torch.nn as nn
-import torchvision.models as models
 
 from qreps.feature_functions import (
     AbstractFeatureFunction,
     AbstractStateActionFeatureFunction,
 )
 from qreps.feature_functions.identity import IdentityFeature
-from torchvision import transforms
+
 from torch.nn import AvgPool2d, Conv2d
 import numpy as np
 
