diff --git a/lightning_research_framework/research/algs/__pycache__/base.cpython-311.pyc b/lightning_research_framework/research/algs/__pycache__/base.cpython-311.pyc
index 55ab614..71ec901 100644
Binary files a/lightning_research_framework/research/algs/__pycache__/base.cpython-311.pyc and b/lightning_research_framework/research/algs/__pycache__/base.cpython-311.pyc differ
diff --git a/lightning_research_framework/research/algs/__pycache__/gumbel_sac_v2.cpython-311.pyc b/lightning_research_framework/research/algs/__pycache__/gumbel_sac_v2.cpython-311.pyc
index d78d652..918ae25 100644
Binary files a/lightning_research_framework/research/algs/__pycache__/gumbel_sac_v2.cpython-311.pyc and b/lightning_research_framework/research/algs/__pycache__/gumbel_sac_v2.cpython-311.pyc differ
diff --git a/lightning_research_framework/research/algs/base.py b/lightning_research_framework/research/algs/base.py
index 9594abf..d4b14a5 100644
--- a/lightning_research_framework/research/algs/base.py
+++ b/lightning_research_framework/research/algs/base.py
@@ -233,7 +233,7 @@ class Algorithm(ABC):
         # Setup profiling immediately before we start the loop.
         start_time = current_time = time.time()
         profiling_metric_lists = defaultdict(list)
-        
+        stop = False
         while current_step < total_steps:
 
             for batch in dataloader:
@@ -318,6 +318,10 @@ class Algorithm(ABC):
                     # Evaluation episodes
                     if self.eval_env is not None and eval_ep > 0:
                         eval_metrics = eval_policy(self.eval_env, self, eval_ep)
+                        if eval_metrics['reward'] < 100 and current_step >= 1000:
+                            print("Stopping early due to poor performance")
+                            stop = True
+                            break
                         if loss_metric in eval_metrics:
                             current_validation_metric = eval_metrics[loss_metric]
                         log_from_dict(logger, eval_metrics, "eval")
@@ -345,6 +349,7 @@ class Algorithm(ABC):
                     break
                 
             self._epochs += 1
+            if stop: break
         logger.close()
 
     @abstractmethod
diff --git a/lightning_research_framework/scripts/run.sh b/lightning_research_framework/scripts/run.sh
index 5d3abbf..6253978 100644
--- a/lightning_research_framework/scripts/run.sh
+++ b/lightning_research_framework/scripts/run.sh
@@ -2,24 +2,15 @@
 
 # List of YAML files
 yaml_files=(
-
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_5.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_6.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_7.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_10.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_11.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_12.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_13.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_14.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_16.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_17.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_18.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_19.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_20.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_21.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_22.yaml"
-    "/content/tests/lightning_research_framework/configs/configs_beta/test_23.yaml"
-
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_13.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_14.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_15.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_16.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_19.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_20.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_21.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_22.yaml"
+    "/Users/nicolasvila/workplace/uni/tfg/tests/lightning_research_framework/configs/configs_beta/test_23.yaml"
 )
 
 # Loop through the YAML files and execute the command
